dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs

```diff
@@ -0,0 +1,265 @@
+    /// </summary>
+    public ProcessStepBuilder(string? name = null)
+        : base(name ?? typeof(TStep).Name)
+    {
+        this._eventNamespace = $"{this.Name}_{this.Id}";
+        this.FunctionsDict = this.GetFuctionMetadataMap();
+    }
+
+    /// <summary>
+    /// Builds the step.
+    /// </summary>
+    /// <returns>An instance of <see cref="KernelProcessStepInfo"/></returns>
+    internal override KernelProcessStepInfo BuildStep()
+    {
+        KernelProcessStepState? stateObject = null;
+
+        if (TryGetSubtypeOfStatefulStep(typeof(TStep), out Type? genericStepType) && genericStepType is not null)
+        {
+            // The step is a subclass of KernelProcessStep<>, so we need to extract the generic type argument
+            // and create an instance of the corresponding KernelProcessStepState<>.
+            var userStateType = genericStepType.GetGenericArguments()[0];
+            Verify.NotNull(userStateType);
+
+            var stateType = typeof(KernelProcessStepState<>).MakeGenericType(userStateType);
+            Verify.NotNull(stateType);
+
+            stateObject = (KernelProcessStepState?)Activator.CreateInstance(stateType, this.Id, this.Name);
+        }
+        else
+        {
+            // The step is a KernelProcessStep with no user-defined state, so we can use the base KernelProcessStepState.
+            stateObject = new KernelProcessStepState(this.Id, this.Name);
+        }
+
+        Verify.NotNull(stateObject);
+
+        // Build the edges first
+        var builtEdges = this.Edges.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Select(e => e.Build()).ToList());
+
+        // Then build the step with the edges and state.
+        var builtStep = new KernelProcessStepInfo(typeof(TStep), stateObject, builtEdges);
+        return builtStep;
+    }
+
+    /// <inheritdoc/>
+    internal override string GetScopedEventId(string eventId)
+    {
+        // Scope the event to this instance of this step by prefixing the event Id with the step's namespace.
+        return $"{this._eventNamespace}.{eventId}";
+    }
+
+    /// <inheritdoc/>
+    internal override Dictionary<string, KernelFunctionMetadata> GetFuctionMetadataMap()
+    {
+        // TODO: Should not have to create a new instance of the step to get the functions metadata.
+        var functions = KernelPluginFactory.CreateFromType<TStep>();
+        return functions.ToDictionary(f => f.Name, f => f.Metadata);
+    }
+
+    /// <summary>
+    /// Attempts to find an instance of <![CDATA['KernelProcessStep<>']]> within the provided types hierarchy.
+    /// </summary>
+    /// <param name="type">The type to examine.</param>
+    /// <param name="genericStateType">The matching type if found, otherwise null.</param>
+    /// <returns>True if a match is found, false otherwise.</returns>
+    private static bool TryGetSubtypeOfStatefulStep(Type? type, out Type? genericStateType)
```

Comment:
I might consider locating this in an external static helper or extension method...just for ease of testing all the arcs.